src/Components/Web/src/Virtualization/Virtualize.cs

```diff
@@ -0,0 +1,304 @@
+
+            var itemsAfter = Math.Max(0, _itemCount - _visibleItemCapacity - _itemsBefore);
+
+            builder.OpenElement(6, "div");
+            builder.AddAttribute(7, "style", GetSpacerStyle(itemsAfter));
+            builder.AddElementReferenceCapture(8, elementReference => _spacerAfter = elementReference);
+
+            builder.CloseElement();
+        }
+
+        private string GetSpacerStyle(int itemsInSpacer)
+            => $"height: {itemsInSpacer * ItemSize}px;";
+
+        void IVirtualizeJsCallbacks.OnBeforeSpacerVisible(float spacerSize, float containerSize)
+        {
+            CalcualteItemDistribution(spacerSize, containerSize, out var itemsBefore, out var visibleItemCapacity);
+
+            UpdateItemDistribution(itemsBefore, visibleItemCapacity);
+        }
+
+        void IVirtualizeJsCallbacks.OnAfterSpacerVisible(float spacerSize, float containerSize)
+        {
+            CalcualteItemDistribution(spacerSize, containerSize, out var itemsAfter, out var visibleItemCapacity);
+
+            var itemsBefore = Math.Max(0, _itemCount - itemsAfter - visibleItemCapacity);
+
+            UpdateItemDistribution(itemsBefore, visibleItemCapacity);
+        }
+
+        private void CalcualteItemDistribution(float spacerSize, float containerSize, out int itemsInSpacer, out int visibleItemCapacity)
+        {
+            itemsInSpacer = Math.Max(0, (int)Math.Floor(spacerSize / ItemSize) - 1);
+            visibleItemCapacity = (int)Math.Ceiling(containerSize / ItemSize) + 2;
+        }
+
+        private void UpdateItemDistribution(int itemsBefore, int visibleItemCapacity)
+        {
+            if (itemsBefore != _itemsBefore || visibleItemCapacity != _visibleItemCapacity)
+            {
+                _itemsBefore = itemsBefore;
+                _visibleItemCapacity = visibleItemCapacity;
+                _refreshTask = RefreshDataAsync();
+
+                if (!_refreshTask.IsCompleted)
+                {
+                    StateHasChanged();
+                }
+            }
+        }
+
+        private async Task RefreshDataAsync()
+        {
+            _refreshCts?.Cancel();
+            _refreshCts = new CancellationTokenSource();
+
+            var cancellationToken = _refreshCts.Token;
+            var request = new ItemsProviderRequest(_itemsBefore, _visibleItemCapacity, cancellationToken);
+
+            try
+            {
+                var result = await _itemsProvider(request);
+
+                // Only apply result if the task was not canceled.
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    _itemCount = result.TotalItemCount;
+                    _loadedItems = result.Items;
+                    _loadedItemsStartIndex = request.StartIndex;
+
+                    StateHasChanged();
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // Bubble-up the cancellation.
```

Comment:
Can you write a note about why it's safe to bubble this up?

Reply:
Your comment made me realize that it's not necessary to bubble-up exceptions anymore, since previous tasks are not `await`ed after they are canceled. Do you think it would be fine to just change this to:

```
try
{
    // ...
}
catch (Exception e) when (!(e is OperationCanceledException))
{
    // Cache this exception so the renderer can throw it.
    _refreshException = e;

    // Re-render the component to throw the exception.
    StateHasChanged();
}
```
?

Reply:
I'd keep this structure so you could add a comment. 

```
catch (OperationCanceledException) when (cts.IsCanceled)
{
  // We canceled this operation so it's fine to suppress these exceptions
}
catch (Exception ex)
{
  ...
}
```

Reply:
Not sure if you’re using an exception filter for reals but if so please don’t :) They don’t work well with AoT compilation.

Reply:
How about this:
```
catch (Exception e)
{
    if (e is OperationCanceledException oce && oce.CancellationToken == cancellationToken)
    {
        // No-op; we canceled the operation, so it's fine to suppress this exception.
    }
    else
    {
        // Cache this exception so the renderer can throw it.
        _refreshException = e;

        // Re-render the component to throw the exception.
        StateHasChanged();
    }
}
```

Reply:
Looks good!